Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XlgDebugger/XlgDebuggerState.cs

```diff
@@ -0,0 +1,386 @@
+        }
+
+        private string GetPipExecutionLevel(PipId pipId) => GetPipExecutionLevel(Analyzer.TryGetPipExePerf(pipId));
+
+        private string GetPipExecutionLevel(PipExecutionPerformance exePerf)
+        {
+            return exePerf == null
+                ? ExeLevelNotCompleted
+                : exePerf.ExecutionLevel.ToString();
+        }
+
+        #region Expression Evaluator
+
+        /// <inheritdoc />
+        public Possible<ObjectContext, Failure> GetCompletions(ThreadState threadState, int frameIndex, string expr)
+        {
+            return new Failure<string>("hi");
+        }
+
+        /// <inheritdoc />
+        public Possible<ObjectContext, Failure> EvaluateExpression(ThreadState threadState, int frameIndex, string expr, bool evaluateForCompletions)
+        {
+            if (Pip.TryParseSemiStableHash(expr, out var hash) && Analyzer.SemiStableHash2Pip.TryGetValue(hash, out var pipId))
+            {
+                return new ObjectContext(context: this, Analyzer.AsPipReference(pipId));
+            }
+
+            string translatedPath = Analyzer.Session.TranslateUserPath(expr);
+            if (AbsolutePath.TryCreate(PathTable, translatedPath, out var path) && path.IsValid)
+            {
+                return new ObjectContext(context: this, new AnalyzePath(path));
+            }
+
+            var exprToEval = evaluateForCompletions ? $"({expr})[0]" : expr;
+            var maybeResult = JPath.JPath.TryEval(Evaluator, exprToEval);
+            if (maybeResult.Succeeded)
+            {
+                return new ObjectContext(this, maybeResult.Result);
+            }
+            else
+            {
+                return maybeResult.Failure;
+            }
+        }
+
+        #endregion
+
+        /// <inheritdoc />
+        public override string ThreadName() => "Full XLG";
+
+        /// <inheritdoc />
+        public override IEnumerable<ObjectContext> GetSupportedScopes(int frameIndex)
+        {
+            return new[]
+            {
+                Renderer.GetObjectInfo(this, Evaluator.TopEnv.Root).WithPreview("Root"),
+                new ObjectInfo("Funs", properties: RootEnv.Vars.Select(VarToProp)),
+                new ObjectInfo("Vars", properties: Evaluator.TopEnv.Vars.Select(VarToProp))
+            }
+            .Select(obj => new ObjectContext(this, obj));
+
+            Property VarToProp(KeyValuePair<string, Result> kvp) => new Property(kvp.Key, kvp.Value);
+        }
+
+        /// <nodoc />
+        public ObjectInfo Render(Renderer renderer, object ctx, object obj)
+        {
+
+            switch (obj)
+            {
+                case AbsolutePath p:               return new ObjectInfo(p.ToString(PathTable));
+                case Result result:                return renderer.GetObjectInfo(ctx, result.IsScalar ? result.Value.First() : result.Value.ToArray());
+                case Function func:                return new ObjectInfo("fun");

```

Comment:
Nit: Should these be all on new lines like how the "return null" is on a new line on line 207

Reply:
I like it this way better.  And this is what I used in other switch statements.